File: proj.lua/actLua/knight/KnightTitleController.lua
require "externCOK2"
require "commonCOK2"
require ("knight.KnightTitleInfo")

KnightTitleController = class("KnightTitleController")
KnightTitleController.__index = KnightTitleController

local _instance = nil

function KnightTitleController:getInstance()
	if nil == _instance then
		_instance = KnightTitleController:create()
	end
	return _instance
end

function KnightTitleController:create()
	local ret = KnightTitleController.new()
	ret:init()
	return ret
end

function KnightTitleController:purgeData()
	_instance = nil
end

function KnightTitleController:init()
	print("init knight xml start")
    self.KnightTitleInfoMap = {}
    self.m_curKnightTitleId = -1
    self.KnightEffMap = {}
    self.initState = 1

	local dict = LocalController:shared():DBXMLManager():getGroupByKey("knight_title");
	local ref = tolua.cast(dict,"cc.Ref")
	local groupDict = LuaController:getInstance():getDicLuaData(ref)

	print("init knight xml start1")

    if nil ~= groupDict then
    	for key, value in pairs(groupDict) do 
    		print("key::::::"..tostring(key))
    		self.KnightTitleInfoMap[ key ] = KnightTitleInfo:create( value )
    	end
    end

    print("init knight xml end")
end

function KnightTitleController:isInit()
	return self.initState
end

-- 后台返回初始化数据
function KnightTitleController:retInitData(dict)
    local ref = tolua.cast(dict,"cc.Ref")
    local params = LuaController:getInstance():getDicLuaData(ref)
    
    for i,v in ipairs(self.KnightTitleInfoMap) do
        v.unlock = 0
    end
    self.m_curKnightTitleId = -1;

    if params["unlock_title"] ~= nil then
        local unlockTable = string.split(params["unlock_title"],",")
        for k,v in pairs(unlockTable) do
            print("KnightTitleController date  id :::  "..tostring(v))
            self.KnightTitleInfoMap[v].unlock = 1
        end
    end
    
    if params["activity"] ~= nil then
        print("KnightTitleController 1  id :::  "..tostring(params["activity"]))
        self.m_curKnightTitleId = params["activity"]
    end
    self:MakeEffectValue()
    self.initState = 2
end

function KnightTitleController:MakeEffectValue()
    self.KnightEffMap = {}
    if nil ~= self.KnightTitleInfoMap[self.m_curKnightTitleId] then
        local kInfo = self.KnightTitleInfoMap[self.m_curKnightTitleId]
        for i=1,20 do
            if nil ~= kInfo.values[i] and nil ~= kInfo.parasMap[i] then
                local cnt = table.getn(kInfo.parasMap[i])
                for j=1,cnt do
                    local eNum = kInfo.parasMap[i][j]
                    if nil == self.KnightEffMap[ eNum ] then
                        self.KnightEffMap[ eNum ] = 0
                    end
                    self.KnightEffMap[ eNum ] = kInfo.values[i]+self.KnightEffMap[ eNum ]
                end
            end
        end

        GlobalData:shared():getPlayerInfo().knight_title = kInfo.power
        -- UIComponent::getInstance()->setUserData(); 改为发消息更新
    end
end

-- 变更称号
function KnightTitleController:startPutOnKnightTitle(itemId)
    local ccdict = CCDictionary:create()
    ccdict:setObject(CCString:create(itemId),"knightId")
    LuaController:getInstance():sendCMD("knight.puton.command",ccdict)
    return true
end

function KnightTitleController:retPutOnKnightTitle(params)
    local ref = tolua.cast(params,"cc.Ref")
    local dict = LuaController:getInstance():getDicLuaData(ref)
    if nil ~= dict["errorCode"] then
        local errorCode = dict["errorCode"]
        CCCommonUtilsForLua:flyText(LuaController:getLang(errorCode))
    elseif nil ~= dict["CurKnightTitle"] then
        print("put on knight title  id :::  "..tostring(dict["CurKnightTitle"]))
        self.m_curKnightTitleId = dict["CurKnightTitle"]
        self:MakeEffectValue()
    end
end

-- 解锁称号
function KnightTitleController:startUnlockKnightTitle(itemId)
    local ccdict = CCDictionary:create()
    ccdict:setObject(CCString:create(itemId),"knightId")
    LuaController:getInstance():sendCMD("knight.unlock.command",ccdict)
    return true
end

function KnightTitleController:retUnlockKnightTitle(params)
    local ref = tolua.cast(params,"cc.Ref")
    local dict = LuaController:getInstance():getDicLuaData(ref)
    if nil ~= dict["errorCode"] then
        local errorCode = dict["errorCode"]
        CCCommonUtilsForLua:flyText(LuaController:getLang(errorCode))
    else
        print("unlock knight knight_title ")
    end
end

function KnightTitleController:UpdateUnlockKnight(params)
    local buildLv = 0;
    local ref = tolua.cast(params,"cc.Ref")
    local dict = LuaController:getInstance():getDicLuaData(ref)
    if nil ~= dict["level"] then
        buildLv = dict["level"]
    end

    if buildLv == 0 then
        return
    end

    for i,v in ipairs(self.KnightTitleInfoMap) do
        if v.unlock == 0 then
            if table.getn(v.mateUnlockMap)==0 and v.bd_level<=buildLv then
                v.unlock = 1
            end
        end
    end

end

-- 合成徽章
function KnightTitleController:startComposeBedge(itemId)
    local ccdict = CCDictionary:create()
    ccdict:setObject(CCString:create(itemId),"knight")
    LuaController:getInstance():sendCMD("material.create",ccdict)
    return true
end
function KnightTitleController:retComposeBedge(dict)
    
end

-- 分解徽章
function KnightTitleController:startDecBedge(itemId)
    local ccdict = CCDictionary:create()
    ccdict:setObject(CCString:create(itemId),"itemId")
    LuaController:getInstance():sendCMD("material.decompse",ccdict)
    return true
end

function KnightTitleController:retDecBedge(dict)
    
end

-- 获取已经使用的徽章数量
function KnightTitleController:GetUseBedgeById(itemId)
    if self.m_curKnightTitleId > 0 then
        if nil ~= self.KnightTitleInfoMap[m_curKnightTitleId].mateUseMap[itemId] then
            return self.KnightTitleInfoMap[m_curKnightTitleId].mateUseMap[itemId]
        end
    else
        return 0
    end
end

File: IF/Classes/controller/KnightTitleController.cpp
//
//  KnightTitleController.cpp
//  IF
//
//  Created by 付彬 on 16/1/7.
//
//

#include "KnightTitleController.hpp"
#include "UIComponent.h"
#include "PopupViewController.h"
#include "LocalController.h"
#include "YesNoDialog.h"
#include "ToolController.h"
#include "FunBuildController.h"
#include "FunBuild.h"
#include "KnightTitleCommand.hpp"
#include "EquipmentCommand.h"
#include "LuaController.h"

static KnightTitleController *_instance = NULL;

KnightTitleController::KnightTitleController()
{
//    auto allDictInfo = LocalController::shared()->DBXMLManager()->getGroupByKey("equipment");
//    if(allDictInfo)
//    {
//        KnightTitleInfoMap.clear();
//        
//        CCDictElement* element;
//        CCDICT_FOREACH(allDictInfo, element)
//        {
//            string strId = element->getStrKey();
//            CCDictionary* dictInfo = _dict(element->getObject());
//            int id = atoi(strId.c_str());
//            KnightTitleInfoMap[id] = KnightTitleInfo(dictInfo);
//            m_KnightLockMap[id] = 0;
//        }
//    }
    CCLOG("init knight xml");
}

KnightTitleController::~KnightTitleController() {}

KnightTitleController* KnightTitleController::getInstance() {
    if (!_instance) {
        _instance = new KnightTitleController();
    }
    return _instance;
}

void KnightTitleController::purgeData() {
    CC_SAFE_RELEASE_NULL( _instance );
    _instance = NULL;
}

//后台返回初始化数据
void KnightTitleController::retInitData(CCDictionary* dict)
{
//    CCDictionary* knightDict = _dict(dict->objectForKey("knight_title"));
//    for (map<int,int>::iterator it = m_KnightLockMap.begin(); it!=m_KnightLockMap.end(); it++) {
//        it->second = 0;
//    }
//    m_curKnightTitleId = -1;
//    
//    string unlock_title = knightDict->valueForKey("unlock_title")->getCString();
//    vector<string> unlockVec;
//    CCCommonUtils::splitString(unlock_title, ",", unlockVec);
//    for (int i=0; i<unlockVec.size(); i++)
//    {
//        int itemId = atoi(unlockVec[i].c_str());
//        m_KnightLockMap[itemId] = 1;
//    }
//    
//    m_curKnightTitleId = knightDict->valueForKey("activity")->intValue();
//    MakeEffectValue();
    
    CCDictionary* knightDict = _dict(dict->objectForKey("knight_title"));
    LuaController::getInstance()->saveLoginData("init_knight", knightDict);
    CCLOG("test");
}

void KnightTitleController::MakeEffectValue()
{
    KnightEffMap.clear();
    if ( KnightTitleInfoMap.find(m_curKnightTitleId) != KnightTitleInfoMap.end() )
    {
        auto& kInfo = KnightTitleInfoMap[m_curKnightTitleId];
        for (int i=1; i<=20; i++)
        {
            if (kInfo.values.find(i) != kInfo.values.end() && kInfo.values[i]>0) {
                for (int j=0; j<kInfo.parasMap[i].size(); j++) {
                    int eNum = kInfo.parasMap[i][j];
                    KnightEffMap[eNum] += kInfo.values[i];
                }
            }
        }
//        int exc = kInfo.power - GlobalData::shared()->playerInfo.knight_title;
        GlobalData::shared()->playerInfo.knight_title = kInfo.power;
        UIComponent::getInstance()->setUserData();
    }
}

//变更称号
bool KnightTitleController::startPutOnKnightTitle(int itemId)
{
    KnightTitlePutOnCommand* cmd = new KnightTitlePutOnCommand(itemId);
    cmd->sendAndRelease();
    return true;
}

void KnightTitleController::retPutOnKnightTitle(CCDictionary* dict)
{
    LuaController::getInstance()->backCMD("knight_puton", dict);
//    if(dict->objectForKey("errorCode"))
//    {
//        string error = dict->valueForKey("errorCode")->getCString();
//        CCCommonUtils::flyText(_lang(error.c_str()));
//    }
//    else
//    {
//        CCLOG("test");
//        m_curKnightTitleId = dict->valueForKey("CurKnightTitle")->intValue();
//        MakeEffectValue();
//    }
}

//解锁称号
bool KnightTitleController::startUnlockKnightTitle(int itemId)
{
    KnightTitleUnlockCommand* cmd = new KnightTitleUnlockCommand(itemId);
    cmd->sendAndRelease();
    return true;
}

void KnightTitleController::retUnlockKnightTitle(CCDictionary* dict)
{
    LuaController::getInstance()->backCMD("knight_unlock", dict);
//    if(dict->objectForKey("errorCode"))
//    {
//        string error = dict->valueForKey("errorCode")->getCString();
//        CCCommonUtils::flyText(_lang(error.c_str()));
//    }
//    else
//    {
//        CCLOG("test");
//    }
}

void KnightTitleController::UpdateUnlockKnight()
{
    int buildLv = 0;
    int bid = FunBuildController::getInstance()->getMaxLvBuildByType(FUN_BUILD_KNIGHT);
    if (bid > 0) {
        auto& bInfo = FunBuildController::getInstance()->getFunbuildById(bid);
        buildLv = bInfo.level;
    }
    if (buildLv == 0) {
        return;
    }
    auto dict = CCDictionary::create();
    dict->setObject(CCString::create(CC_ITOA(buildLv)), "level");
    LuaController::getInstance()->backCMD("knight_update", dict);
//    map<int, int>::iterator it = m_KnightLockMap.begin();
//    for (; it!=m_KnightLockMap.end(); it++)
//    {
//        if (it->second == 0)
//        {
//            auto& kInfo = KnightTitleInfoMap[it->first];
//            if (kInfo.mateUnlockMap.size() <= 0)
//            {
//                if (kInfo.bd_level <= buildLv) {
//                    it->second = 1;
//                }
//            }
//        }
//    }
}

//合成徽章
bool KnightTitleController::startComposeBedge(int itemId)
{
    MaterialCreateCommand* cmd = new MaterialCreateCommand(CC_ITOA(itemId), 3);
    cmd->sendAndRelease();
    return true;
}
void KnightTitleController::retComposeBedge(CCDictionary* dict)
{
    if(dict->objectForKey("errorCode"))
    {
        string error = dict->valueForKey("errorCode")->getCString();
        if(error=="102198"){//不显示材料不足弹出提示
            return;
        }
        CCCommonUtils::flyText(_lang(error.c_str()));
    }
    else
    {
        CCLOG("test");
        if (dict->objectForKey("dec")) {//消耗掉的道具信息
            auto decDic = _dict(dict->objectForKey("dec"));
            ToolController::getInstance()->pushAddTool(decDic);
        }
        if (dict->objectForKey("add")) {//新增的道具信息
            auto addDic = _dict(dict->objectForKey("add"));
            ToolController::getInstance()->pushAddTool(addDic);
        }
//        CCSafeNotificationCenter::sharedNotificationCenter()->postNotification(MSG_MATE_CRT_END);
    }

}
//分解徽章
bool KnightTitleController::startDecBedge(int itemId)
{
    MaterialDecomposeCommand* cmd = new MaterialDecomposeCommand(CC_ITOA(itemId));
    cmd->sendAndRelease();
    return true;
}
void KnightTitleController::retDecBedge(CCDictionary* dict)
{
    if(dict->objectForKey("errorCode"))
    {
        string error = dict->valueForKey("errorCode")->getCString();
        CCCommonUtils::flyText(_lang(error.c_str()));
    }
    else
    {
        CCLOG("test");
        if (dict->objectForKey("dec")) {//消耗掉的道具信息
            auto decDic = _dict(dict->objectForKey("dec"));
            ToolController::getInstance()->pushAddTool(decDic);
        }
        if (dict->objectForKey("add")) {//新增的道具信息
            auto addDic = _dict(dict->objectForKey("add"));
            ToolController::getInstance()->pushAddTool(addDic);
        }
    }
}

//获取已经使用的徽章数量
int KnightTitleController::GetUseBedgeById(int itemId)
{
    if (m_curKnightTitleId > 0)
    {
        if ( KnightTitleInfoMap[m_curKnightTitleId].mateUseMap.find(itemId) != KnightTitleInfoMap[m_curKnightTitleId].mateUseMap.end())
        {
            return KnightTitleInfoMap[m_curKnightTitleId].mateUseMap[itemId];
        }
    }
    else
    {
        return 0;
    }
}

